src/hotspot/share/utilities/checkedCast.hpp

```diff
@@ -25,22 +25,177 @@
+                         std::conditional_t<to_signed, ToSigned, FromSigned>>;
+  };
+
+public:
+  template<typename To, typename From>
+  static constexpr bool is_tautology() {
+    using Dispatcher = Dispatcher<To, From, true>;
+    using Checker = typename Dispatcher::type;
+    return Checker::is_tautology;
+  }
+
+  template<typename To, bool permit_tautology, typename From>
+  static constexpr bool check(From from) {
+    using Dispatcher = Dispatcher<To, From, permit_tautology>;
+    using Checker = typename Dispatcher::type;
+    return Checker()(from);
+  }
+};
+
+// If both types are signed or both are unsigned, only a narrowing conversion
+// can lose information.  We check for such loss via a round-trip conversion.
+// C++14 4.7 defines the result for unsigned types, and implementation-defined
+// for signed types.  All supported implementations do the "obvious" discard
+// of high-order bits when narrowing signed values, and C++20 requires that
+// behavior.  This approach avoids any additional complexity to avoid
+// potential tautological comparisons.
+template<typename To, typename From, bool permit_tautology>
+struct CheckedCastImpl::CheckSameSign {
+  static const bool is_narrowing = sizeof(To) < sizeof(From);
+  static const bool is_tautology = !is_narrowing;
+  static_assert(permit_tautology || is_narrowing, "tautological checked_cast");
+  constexpr bool operator()(From from) const {
+    return !is_narrowing || (static_cast<From>(static_cast<To>(from)) == from);
+  }
+};
+
+// Conversion from unsigned to signed is okay if value <= To's max.
+template<typename To, typename From, bool permit_tautology>
+struct CheckedCastImpl::CheckToSigned<To, From, permit_tautology, true /* is_narrowing */> {
+  static const bool is_tautology = false;
+  constexpr bool operator()(From from) const {
+    return check_range<To>(from);
+  }
+};
+
+// Avoid tautological comparison when not narrowing.
+template<typename To, typename From, bool permit_tautology>
+struct CheckedCastImpl::CheckToSigned<To, From, permit_tautology, false /* is_narrowing */> {
+  static const bool is_tautology = true;
+  static_assert(permit_tautology, "tautological checked_cast");
+  constexpr bool operator()(From from) const {
+    return true;
+  }
+};
+
+// Conversion from signed to unsigned is okay when 0 <= value <= To's max.
+template<typename To, typename From, bool permit_tautology>
+struct CheckedCastImpl::CheckFromSigned<To, From, permit_tautology, true /* is_narrowing */> {
+  static const bool is_tautology = false;
+  constexpr bool operator()(From from) const {
+    return (from >= 0) && check_range<To>(from);
+  }
+};
+
+// Avoid tautological comparison when not narrowing.
+template<typename To, typename From, bool permit_tautology>
+struct CheckedCastImpl::CheckFromSigned<To, From, permit_tautology, false /* is_narrowing */> {
+  static const bool is_tautology = false;
+  constexpr bool operator()(From from) const {
+    return (from >= 0);
+  }
+};
```

Comment:
So instead of checked_cast<uint64_t, int32_t> or checked_cast<uint,int>, callers should check from < 0 and then do a static_cast<>?  It might be better to allow checked_cast to support non-narrowed signed to unsigned, because I think the alternative is more error-prone.